src/vs/workbench/contrib/terminal/browser/terminalInstance.ts

```diff
@@ -967,25 +966,13 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
 				this._messageTitleDisposable = this._processManager.onProcessTitle(title => this.setTitle(title ? title : '', TitleEventSource.Process));
 			}
 		}
-
-		if (platform.isWindows) {
-			this._processManager.ptyProcessReady.then(() => {
-				if (this._processManager.remoteAuthority) {
-					return;
-				}
-				this._xtermReadyPromise.then(xterm => {
-					if (!this._isDisposed && this._processManager && this._processManager.shellProcessId) {
-						this._windowsShellHelper = this._terminalInstanceService.createWindowsShellHelper(this._processManager.shellProcessId, xterm);
-						this._windowsShellHelper.onShellNameChange(title => {
-							this.setShellType(this.getShellType(title));
-							if (this.isTitleSetByProcess && !this._configHelper.config.experimentalUseTitleEvent) {
-								this.setTitle(title, TitleEventSource.Process);
-							}
-						});
-					}
-				});
-			});
-		}
+		this._processManager.onShellTypeChanged(type => {
+			// only set if it's changed
+			if (this._shellType !== type) {
+				this.setShellType(type);
+				this.setTitle(type, TitleEventSource.Process);
+			}
```

Comment:
Instead of calling `setTitle` using the shell type, let's fire `onProcessTitleChanged` on TerminalProcess, that way it gets handles by the Linux/macOS mechanism.

Reply:
I can't pass type of `TerminalShellType` to `onProcessTitle` of `ProcessManager`

Reply:
Oh I think we might already be doing this here:

https://github.com/microsoft/vscode/blob/d0a514a651e52fcb3c4ab1619fe85523f67bd0ee/src/vs/platform/terminal/node/terminalProcess.ts#L138-L139

Something that could be nice to clean that up a bit further is exposing both title and shell type events on `IWindowsShellHelper`, that way we not export `getShellType` and to avoid `TerminalProcess` knowing anything about it. Then hook it up like:

```ts
this._register(this._windowsShellHelper.onShellTypeChanged(e=> this._onProcessShellTypeChanged.fire(e)));
this._register(this._windowsShellHelper.onTitleChanged(e=> this._onProcessTitleChanged.fire(e)));
```